### PR TITLE
Fixing HC_CALL memory management

### PR DESCRIPTION
**Pull request by jasonsandlin**

### CODE REVIEW

File: Source/HTTP/httpcall.cpp [hunk 1]
```diff
@@ -371,7 +374,8 @@ void retry_http_call_until_done(
     auto httpSingleton = get_http_singleton(false);
     if (nullptr == httpSingleton)
     {
-        XAsyncComplete(retryContext->outerAsyncBlock, S_OK, 0);
+        HC_TRACE_WARNING(HTTPCLIENT, "Http call after HCCleanup was called. Aborting call.");
```
Comment (jplafonta) [inline Source/HTTP/httpcall.cpp]: > HC_TRACE_WARNING(HTTPCLIENT, "Http call after HCCleanup was called. Aborting call."); [](start = 7, length = 86)

Does HC_TRACE_WARNING rely on the global state?

File: Source/Common/utils.cpp [hunk 2]
```diff
@@ -82,16 +82,20 @@ HRESULT RunAsync(
     uint64_t delayInMs
 )
 {
-    auto workPointer = new AsyncWork(std::move(work));
+    auto workPointer = Make<AsyncWork>(std::move(work));
```
Comment (lubeltra) [inline Source/Common/utils.cpp]: The pattern I would suggest for all this kind of code:
```C++
// start with the async block and/or context being owned by a smart pointer
// if you allocate a context object, just put the async block inside it, it saves you an allocation
std::unique_ptr<XAsyncBlock> async = Make<XAsyncBlock>();
if (!async) { return E_OUTOFMEMORY; }
async->callback = [](XAsyncBlock* asyncPtr)
{
    std::unique_ptr<XAsyncBloc> async{ asyncPtr };
    // do callback
};

// blah blah more setup

HRESULT hr = XAsyncBegin(
    async.get(), // don't release yet! we could fail still
    contextPtrIfAny,
    stuff,
    "stuff",
    [](XAsyncOp op, const XAsyncProviderData* data)
    {
        switch (op)
        {
            case XAsyncOp::Begin:
                return XAsyncSchedule(data->async); // in this case you need to know the delay, you can pass it in through your context object
            case XAsyncOp::DoWork:
                // do work
        }
});
if (FAILED(hr)) { return hr; }
async.release(); // at this point we know do work will be called eventually

Comment (lubeltra) [reply Source/Common/utils.cpp]: You'll want a typedef of std::unique_ptr with a custom deleter for this

File: Source/HTTP/httpcall.cpp (hunk 1 above)
Comment (jasonsandlin) [reply Source/HTTP/httpcall.cpp]: No, it uses

TraceState& GetTraceState() noexcept
{
    static TraceState state;
    return state;
}

and bails out if it that's not setup

File: Source/Common/utils.cpp [hunk 3]
```diff
@@ -82,25 +82,25 @@ HRESULT RunAsync(
...
 
     asyncOp->queue = queue;
-    asyncOp->context = workPointer;
+    asyncOp->context = workPointer.get();
     asyncOp->callback = [](XAsyncBlock* asyncOp)
     {
-        auto context = static_cast<AsyncWork*>(asyncOp->context);
-        Delete(context);
-        Delete(asyncOp);
+        HC_UNIQUE_PTR<AsyncWork> context{ reinterpret_cast<AsyncWork*>(asyncOp->context) };
```
Comment (lubeltra) [inline Source/Common/utils.cpp]: `static_cast` should be sufficient

File: Source/HTTP/httpcall.cpp [hunk 4]
```diff
@@ -401,57 +405,64 @@ void retry_http_call_until_done(
         }
     }
 
+    XAsyncBlock* nestedBlock = Make<XAsyncBlock>();
```
Comment (lubeltra) [inline Source/HTTP/httpcall.cpp]: you should do the same changes here

Comment (jasonsandlin) [reply Source/HTTP/httpcall.cpp]: done

**Pull request merged**